Fix accidental ignition and stuck graceful termination

tnozicka opened the pull request:
**Description of your changes:**
This PR fixes the ignition touching the file before it's actually ignited and adds e2e coverage to prevent it from happening again. (Looks like it got broken in the early iterations.)

It also fixes stuck graceful termination *before* ignition is done because fixing one surfaces the other is broken and they can't be split without disabling some tests. (For this reason the ignition file is changed twice - best viewed as an overall diff then "by commit"). It also fixes graceful termination *after* ignition that was broken because of the manager agent container wasn't using `exec`. (@rzetelskik found that previously and it also failed in the test I wrote.)

**Which issue is resolved by this Pull Request:**
Resolves #2213

tnozicka commented:
(the e2e test on it's own failed as expected - the run was https://prow.scylla-operator.scylladb.com/view/gs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/2212/pull-scylla-operator-master-e2e-gke-serial/1862139318282227712#1:test-build-log.txt%3A994)

tnozicka commented on test/e2e/set/nodeconfig/nodeconfig_optimizations.go:
```diff
@@ -328,6 +329,25 @@ var _ = g.Describe("NodeConfig Optimizations", framework.Serial, func() {
+					o.HaveKeyWithValue(naming.ScyllaContainerName, false),
+					o.HaveKeyWithValue(naming.ScyllaDBIgnitionContainerName, false),
+				),
+				fmt.Sprintf("container in Pod %q don't match the expected state", pod.Name),
```
ok, I suppose `container(s)` as it's 1-N

tnozicka commented:
now that we actually wait for ignition - the tuning took over 9 minutes which made the job timeout
https://gcsweb.scylla-operator.scylladb.com/gcs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/2212/pull-scylla-operator-master-e2e-gke-serial/1862151638882455552/artifacts/e2e/cluster-0/namespaces/e2e-test-nodeconfig-z4wpb-0-j4m2b/pods/basic-8l7bn-us-east-1-us-east-1a-0/scylladb-ignition.current

tnozicka commented:
blocked on https://github.com/scylladb/scylla-operator/issues/2216

tnozicka commented:
I've finally traced the failed serial tests ona GKE cluster - indicating graceful termination issues
```
Dec 04 17:15:13 gke-so-753fb68c-ed81-4a23-aa2-workers-24a06bcc-636c kubelet[38896]: I1204 17:15:13.553585   38896 kuberuntime_container.go:647] "PreStop hook completed" pod="e2e-test-nodeconfig-sf5hl-0-ss4tc/basic-rk6zh-us-east-1-us-east-1a-0" podUID="b01f1990-37ff-4da0-8a50-1b39c3024663" containerName="scylla" containerID="containerd://c3d9dd8656b21c701f5ead2387f30d898a1ebc6a58b5d9edf982f184b28033eb"
Dec 04 17:15:13 gke-so-753fb68c-ed81-4a23-aa2-workers-24a06bcc-636c kubelet[38896]: I1204 17:15:13.553605   38896 kuberuntime_container.go:745] "Killing container with a grace period" pod="e2e-test-nodeconfig-sf5hl-0-ss4tc/basic-rk6zh-us-east-1-us-east-1a-0" podUID="b01f1990-37ff-4da0-8a50-1b39c3024663" containerName="scylla" containerID="containerd://c3d9dd8656b21c701f5ead2387f30d898a1ebc6a58b5d9edf982f184b28033eb" gracePeriod=885
Dec 04 17:29:58 gke-so-753fb68c-ed81-4a23-aa2-workers-24a06bcc-636c kubelet[38896]: I1204 17:29:58.725803   38896 kuberuntime_container.go:754] "Container exited normally" pod="e2e-test-nodeconfig-sf5hl-0-ss4tc/basic-rk6zh-us-east-1-us-east-1a-0" podUID="b01f1990-37ff-4da0-8a50-1b39c3024663" containerName="scylla" containerID="containerd://c3d9dd8656b21c701f5ead2387f30d898a1ebc6a58b5d9edf982f184b28033eb"
```

tnozicka commented:
#2246 landed
/test all

zimnx reviewed (commented):
One comment, otherwise lgtm. Please fix PR description, one sentence contains duplicated parts.

tnozicka commented on test/e2e/set/scyllacluster/scyllacluster_restarts.go:
```diff
@@ -0,0 +1,217 @@
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		pod, err = f.KubeClient().CoreV1().Pods(f.Namespace()).Get(ctx, podName, metav1.GetOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
```
fixed, thx

tnozicka commented:
> Please fix PR description, one sentence contains duplicated parts.

I fixed the typos and added more context, but one fix is for graceful termination *before* ignition and the other fix is *after* ignition.

zimnx reviewed (approved):
lgtm
/assign rzetelskik

rzetelskik commented on test/e2e/set/nodeconfig/nodeconfig_optimizations.go:
```diff
@@ -400,7 +421,20 @@ var _ = g.Describe("NodeConfig Optimizations", framework.Serial, func() {
+			scyllaclusterverification.VerifyOptions{
+				VerifyStatefulSetOptions: scyllaclusterverification.VerifyStatefulSetOptions{
+					PodRestartCountAssertion: func(a o.Assertion, containerName, podName string) {
+						// We expect restart(s) from the startup probe, usually 1.
+						a.To(o.BeNumerically("<=", 2), fmt.Sprintf("container %q in pod %q should not be restarted by the startup probe more than twice", containerName, podName))
```
Not sure I understand this right - isn't this testing a side effect? If you lowered the period of the startup probe, or the timeout (which according to [this TODO](https://github.com/scylladb/scylla-operator/blob/d4da9dfe44e55d7dd674a3d0394eafb5f309af2f/pkg/controller/scylladbdatacenter/resource.go#L784) should have already had been done), won't it introduce flakiness?

tnozicka replied:
Nope, my understanding is that this would straight fail and you'd need to adjust the value. Similarly the 10m30s sleep should be adjusted as well.

rzetelskik replied:
ok, let's give this a try

rzetelskik reviewed (commented):
one comment/question, rest lgtm

rzetelskik commented:
/lgtm

tnozicka commented:
bug https://github.com/scylladb/scylla-operator/issues/2249
/retest
